### PR TITLE
Add context variables to OuterLoop

### PR DESCRIPTION
**Pull request by marpulli**
*Issue #, if available:* #15 

*Description of changes:* Add ability to pass in dictionary of context variables to `OuterLoop.run()`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: emukit/core/optimization/acquisition_optimizer.py [hunk 1]
```diff
@@ -28,6 +29,12 @@ def optimize(self, acquisition: Acquisition, context: dict = None) -> Tuple[np.n
         :param context: Optimization context, determines whether any variable values should be fixed during the optimization
         """
 
+        # Check that there are some free parameters to optimize
+        if context is not None:
+            are_all_parameters_fixed = all([name in context.keys() for name in self.space.parameter_names])
```
[the remark this answers is not in the document]
Comment (marpulli) [reply emukit/core/optimization/acquisition_optimizer.py]: correct, but this was a neat one liner!

File: emukit/core/loop/outer_loop.py [hunk 2]
```diff
@@ -61,7 +65,7 @@ def run_loop(self, user_function: Union[UserFunction, Callable], stopping_condit
         while not stopping_condition.should_stop(self.loop_state):
             _log.info("Iteration {}".format(self.loop_state.iteration))
             self.model_updater.update(self.loop_state)
-            new_x = self.candidate_point_calculator.compute_next_points(self.loop_state)
+            new_x = self.candidate_point_calculator.compute_next_points(self.loop_state, context)
```
[the remark this answers is not in the document]
Comment (marpulli) [reply emukit/core/loop/outer_loop.py]: I should check the names and throw an exception if there are context parameters with names we don't recognise. If the value is out of the domain I think a warning is more appropriate.

File: emukit/core/optimization/acquisition_optimizer.py [hunk 3]
```diff
@@ -28,6 +29,12 @@ def optimize(self, acquisition: Acquisition, context: dict = None) -> Tuple[np.n
...
+            if are_all_parameters_fixed:
+                raise ValueError('All parameters are fixed in the context so there is nothing to optimize.')
```
[the remark this answers is not in the document]
Comment (marpulli) [reply emukit/core/optimization/acquisition_optimizer.py]: The optimizer can't run if it has nothing to optimize and would throw an error. However, I can just convert the context to a vector using the context manager then return immediately without calling the optimizer..

File: emukit/core/loop/candidate_point_calculators.py [hunk 4]
```diff
@@ -69,24 +73,33 @@ def _get_information_source_parameter(self) -> InformationSourceParameter:
...
         :param loop_state: Object that tracks the state of the loop. Currently unused
+        :param context: Contains variables to fix through optimization of acquisition function. The dictionary key is
+                        the parameter name and the value is the value to fix the parameter to.
         :return: A list of function inputs to evaluate next
         """
         f_mins = np.zeros((len(self.source_parameter.domain)))
         x_opts = []
 
+        if context is None:
+            context = dict()
```
Comment (apaleyes) [inline emukit/core/loop/candidate_point_calculators.py]: Can we just make this the default value in the method signature?

File: emukit/core/optimization/acquisition_optimizer.py [hunk 5]
```diff
@@ -26,14 +27,38 @@ def optimize(self, acquisition: Acquisition, context: dict = None) -> Tuple[np.n
...
+        :return: Tuple of (optimum value, location of optimum)
         """
 
-        self.gpyopt_acquisition_optimizer.context_manager = GPyOpt.optimization.acquisition_optimizer.ContextManager(
-            self.gpyopt_space, context)
-
         # Take negative of acquisition function because they are to be maximised and the optimizers minimise
         f = lambda x: -acquisition.evaluate(x)
 
+        # Validate context
```
Comment (apaleyes) [inline emukit/core/optimization/acquisition_optimizer.py]: This method becomes a bit lengthy. Can you create smaller functions to move some code in there and reduce the complexity? An easy one: `_validate_context`

File: emukit/core/loop/candidate_point_calculators.py (hunk 4 above)
Comment (marpulli) [reply emukit/core/loop/candidate_point_calculators.py]: I thought having mutable default arguments was dangerous somehow:

https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html

File: emukit/core/optimization/acquisition_optimizer.py (hunk 5 above)
Comment (marpulli) [reply emukit/core/optimization/acquisition_optimizer.py]: Will do

File: emukit/core/loop/candidate_point_calculators.py (hunk 4 above)
Comment (apaleyes) [reply emukit/core/loop/candidate_point_calculators.py]: Wow, I learned something new today. Thanks, this answers

**Pull request merged**